docs/proposals/hpa/federated-hpa-v2.md

```diff
@@ -0,0 +1,785 @@
+    // +optional
+    Status FederatedHPAStatus `json:"status,omitempty"`
+}
+
+type FederatedHPASpec struct {
+    // ScaleTargetRef points to the target resource to scale, and is used to the pods for which metrics
+    // should be collected, as well as to actually change the replica count.
+    ScaleTargetRef autoscalingv2.CrossVersionObjectReference `json:"scaleTargetRef"`
+
+    // Metrics contains the specifications for which to use to calculate the
+    // desired replica count (the maximum replica count across all metrics will
+    // be used).  The desired replica count is calculated multiplying the
+    // ratio between the target value and the current value by the current
+    // number of pods.  Ergo, metrics used must decrease as the pod count is
+    // increased, and vice-versa.  See the individual metric source types for
+    // more information about how each type of metric must respond.
+    // +optional
+    Metrics []autoscalingv2.MetricSpec `json:"metrics"`
+
+    // behavior configures the scaling behavior of the target
+    // in both Up and Down directions (scaleUp and scaleDown fields respectively).
+    // If not set, the default HPAScalingRules for scale up and scale down are used.
+    // +optional
+    Behavior *autoscalingv2.HorizontalPodAutoscalerBehavior `json:"behavior,omitempty"`
+
+    // ClusterAffinity represents scheduling restrictions to a certain set of clusters.
+    // If not set, any cluster can be scheduling candidate.
+    // +optional
+    ClusterAffinity *policyv1alpha1.ClusterAffinity `json:"clusterAffinity,omitempty"`
+
+    // FederatedMinReplicas is the lower limit for the number of replicas to which the federatedHPA controller
+    // can scale down. It defaults to 1 pod and it should not be less than 1.
+    // +optional
+    FederatedMinReplicas *int32 `json:"federatedMinReplicas,omitempty"`
+
+    // FederatedMaxReplicas is the upper limit for the number of replicas to which the federatedHPA controller
+    // can scale up. It cannot be less than FederatedMinReplicas.
+    FederatedMaxReplicas *int32 `json:"federatedMaxReplicas"`
+
+    // AutoscaleMultiClusterDelay is the delay for executing autoscaling action in other clusters once
+    // the current cluster could not scale the workload. It is used to avoid the situation that cluster-autoscaler is used.
+    // +optional
+    AutoscaleMultiClusterDelay *int32 `json:"autoscaleMultiClusterDelay,omitempty"`
+
+    // ScalingToZero indicates whether the workload can be scaled to zero in member clusters.
+    // +optional
+    ScalingToZero bool `json:"scalingToZero,omitempty"`
+
+	// FederatedHPAAssignment is the assignment policy of FederatedMinReplicas/FederatedMaxReplicas
+	// +required
+    FederatedHPAAssignment FederatedHPAAssignment `json:"federatedHPAAssignment,omitempty"`
+}
+
+// FederatedHPAAssignmentType describes assignment methods for FederatedHPA FederatedMinReplicas/FederatedMaxReplicas
+type FederatedHPAAssignmentType string
```

Comment:
I'm curious, is the `assignment` here done by the Karmada scheduler or by the FederatedHP controller?

Reply:
On the other hand, does this field implicitly contain the meaning of "scheduling"? Does it make the user wonder why there are two places where a scheduling policy needs to be set at the same time?

Reply:
> I'm curious, is the assignment here done by the Karmada scheduler or by the FederatedHP controller?

FederatedHPA controller

> On the other hand, does this field implicitly contain the meaning of "scheduling"?

Yes

> Does it make the user wonder why there are two places(PP and FederatedHPA) where a scheduling policy needs to be set at the same time?

Once FederatedHPA is applied, PP is ignored, and the police could be not the same. The reason why we do this is:
1. If we continue to use PP, it will make the FederatedHPA's evolution difficult.
2. PP lacks some policies, and it's not meet the HPA’s requirements fully.

Reply:
Can we implement the scheduling strategy in FederatedHPA by developing a scheduler plugin?


I am worried that implementing scheduling in FederatedHPA will complicate the overall architecture of Karmada.

Reply:
Yeah, this is what I worry too.

From my point of view, the scheduler only cares about scheduling, but FederatedHPA needs other operations, like creating HPA works and updating scaling gvr.

Or FederatedHPA controller could call the scheduler to do assignments?

Reply:
Currently, there is a field defined `Placement` in our RB. Can we transfer the configuration related to scheduling strategy in `FederatedHPA` to RB's Placement through `FederatedHPA-controller`?